File: src/frontend/etiquetas/lib/app/data/model/etiqueta_layout_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:etiquetas/app/data/model/model_imports.dart';

class EtiquetaLayoutModel {
	int? id;
	int? idFormatoPapel;
	String? codigoFabricante;
	int? quantidade;
	int? quantidadeHorizontal;
	int? quantidadeVertical;
	int? margemSuperior;
	int? margemInferior;
	int? margemEsquerda;
	int? margemDireita;
	int? espacamentoHorizontal;
	int? espacamentoVertical;
	List<EtiquetaTemplateModel>? etiquetaTemplateModelList;
	EtiquetaFormatoPapelModel? etiquetaFormatoPapelModel;

	EtiquetaLayoutModel({
		this.id,
		this.idFormatoPapel,
		this.codigoFabricante,
		this.quantidade,
		this.quantidadeHorizontal,
		this.quantidadeVertical,
		this.margemSuperior,
		this.margemInferior,
		this.margemEsquerda,
		this.margemDireita,
		this.espacamentoHorizontal,
		this.espacamentoVertical,
		this.etiquetaTemplateModelList,
		this.etiquetaFormatoPapelModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'codigo_fabricante',
		'quantidade',
		'quantidade_horizontal',
		'quantidade_vertical',
		'margem_superior',
		'margem_inferior',
		'margem_esquerda',
		'margem_direita',
		'espacamento_horizontal',
		'espacamento_vertical',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Codigo Fabricante',
		'Quantidade',
		'Quantidade Horizontal',
		'Quantidade Vertical',
		'Margem Superior',
		'Margem Inferior',
		'Margem Esquerda',
		'Margem Direita',
		'Espacamento Horizontal',
		'Espacamento Vertical',
	];

	EtiquetaLayoutModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idFormatoPapel = jsonData['idFormatoPapel'];
		codigoFabricante = jsonData['codigoFabricante'];
		quantidade = jsonData['quantidade'];
		quantidadeHorizontal = jsonData['quantidadeHorizontal'];
		quantidadeVertical = jsonData['quantidadeVertical'];
		margemSuperior = jsonData['margemSuperior'];
		margemInferior = jsonData['margemInferior'];
		margemEsquerda = jsonData['margemEsquerda'];
		margemDireita = jsonData['margemDireita'];
		espacamentoHorizontal = jsonData['espacamentoHorizontal'];
		espacamentoVertical = jsonData['espacamentoVertical'];
		etiquetaTemplateModelList = (jsonData['etiquetaTemplateModelList'] as Iterable?)?.map((m) => EtiquetaTemplateModel.fromJson(m)).toList() ?? [];
		etiquetaFormatoPapelModel = jsonData['etiquetaFormatoPapelModel'] == null ? EtiquetaFormatoPapelModel() : EtiquetaFormatoPapelModel.fromJson(jsonData['etiquetaFormatoPapelModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idFormatoPapel'] = idFormatoPapel != 0 ? idFormatoPapel : null;
		jsonData['codigoFabricante'] = codigoFabricante;
		jsonData['quantidade'] = quantidade;
		jsonData['quantidadeHorizontal'] = quantidadeHorizontal;
		jsonData['quantidadeVertical'] = quantidadeVertical;
		jsonData['margemSuperior'] = margemSuperior;
		jsonData['margemInferior'] = margemInferior;
		jsonData['margemEsquerda'] = margemEsquerda;
		jsonData['margemDireita'] = margemDireita;
		jsonData['espacamentoHorizontal'] = espacamentoHorizontal;
		jsonData['espacamentoVertical'] = espacamentoVertical;
		
		var etiquetaTemplateModelLocalList = []; 
		for (EtiquetaTemplateModel object in etiquetaTemplateModelList ?? []) { 
			etiquetaTemplateModelLocalList.add(object.toJson); 
		}
		jsonData['etiquetaTemplateModelList'] = etiquetaTemplateModelLocalList;
		jsonData['etiquetaFormatoPapelModel'] = etiquetaFormatoPapelModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idFormatoPapel = plutoRow.cells['idFormatoPapel']?.value;
		codigoFabricante = plutoRow.cells['codigoFabricante']?.value;
		quantidade = plutoRow.cells['quantidade']?.value;
		quantidadeHorizontal = plutoRow.cells['quantidadeHorizontal']?.value;
		quantidadeVertical = plutoRow.cells['quantidadeVertical']?.value;
		margemSuperior = plutoRow.cells['margemSuperior']?.value;
		margemInferior = plutoRow.cells['margemInferior']?.value;
		margemEsquerda = plutoRow.cells['margemEsquerda']?.value;
		margemDireita = plutoRow.cells['margemDireita']?.value;
		espacamentoHorizontal = plutoRow.cells['espacamentoHorizontal']?.value;
		espacamentoVertical = plutoRow.cells['espacamentoVertical']?.value;
		etiquetaTemplateModelList = [];
		etiquetaFormatoPapelModel = EtiquetaFormatoPapelModel();
		etiquetaFormatoPapelModel?.nome = plutoRow.cells['etiquetaFormatoPapelModel']?.value;
	}	

	EtiquetaLayoutModel clone() {
		return EtiquetaLayoutModel(
			id: id,
			idFormatoPapel: idFormatoPapel,
			codigoFabricante: codigoFabricante,
			quantidade: quantidade,
			quantidadeHorizontal: quantidadeHorizontal,
			quantidadeVertical: quantidadeVertical,
			margemSuperior: margemSuperior,
			margemInferior: margemInferior,
			margemEsquerda: margemEsquerda,
			margemDireita: margemDireita,
			espacamentoHorizontal: espacamentoHorizontal,
			espacamentoVertical: espacamentoVertical,
			etiquetaTemplateModelList: etiquetaTemplateModelListClone(etiquetaTemplateModelList!),
		);			
	}

	etiquetaTemplateModelListClone(List<EtiquetaTemplateModel> etiquetaTemplateModelList) { 
		List<EtiquetaTemplateModel> resultList = [];
		for (var etiquetaTemplateModel in etiquetaTemplateModelList) {
			resultList.add(
				EtiquetaTemplateModel(
					id: etiquetaTemplateModel.id,
					idEtiquetaLayout: etiquetaTemplateModel.idEtiquetaLayout,
					tabela: etiquetaTemplateModel.tabela,
					campo: etiquetaTemplateModel.campo,
					formato: etiquetaTemplateModel.formato,
					quantidadeRepeticoes: etiquetaTemplateModel.quantidadeRepeticoes,
					filtro: etiquetaTemplateModel.filtro,
				)
			);
		}
		return resultList;
	}

	
}

File: src/backend/etiquetas/src/model/etiqueta_layout_model.py
from src import db
from src.model.etiqueta_template_model import EtiquetaTemplateModel
from src.model.etiqueta_formato_papel_model import EtiquetaFormatoPapelModel


class EtiquetaLayoutModel(db.Model):
    __tablename__ = 'etiqueta_layout'

    id = db.Column(db.Integer, primary_key=True)
    codigo_fabricante = db.Column(db.String(50))
    quantidade = db.Column(db.Integer)
    quantidade_horizontal = db.Column(db.Integer)
    quantidade_vertical = db.Column(db.Integer)
    margem_superior = db.Column(db.Integer)
    margem_inferior = db.Column(db.Integer)
    margem_esquerda = db.Column(db.Integer)
    margem_direita = db.Column(db.Integer)
    espacamento_horizontal = db.Column(db.Integer)
    espacamento_vertical = db.Column(db.Integer)
    id_formato_papel = db.Column(db.Integer, db.ForeignKey('etiqueta_formato_papel.id'))

    etiqueta_template_model_list = db.relationship('EtiquetaTemplateModel', lazy='dynamic')
    etiqueta_formato_papel_model = db.relationship('EtiquetaFormatoPapelModel', foreign_keys=[id_formato_papel])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_formato_papel = data.get('idFormatoPapel')
        self.codigo_fabricante = data.get('codigoFabricante')
        self.quantidade = data.get('quantidade')
        self.quantidade_horizontal = data.get('quantidadeHorizontal')
        self.quantidade_vertical = data.get('quantidadeVertical')
        self.margem_superior = data.get('margemSuperior')
        self.margem_inferior = data.get('margemInferior')
        self.margem_esquerda = data.get('margemEsquerda')
        self.margem_direita = data.get('margemDireita')
        self.espacamento_horizontal = data.get('espacamentoHorizontal')
        self.espacamento_vertical = data.get('espacamentoVertical')

    def serialize(self):
        return {
            'id': self.id,
            'idFormatoPapel': self.id_formato_papel,
            'codigoFabricante': self.codigo_fabricante,
            'quantidade': self.quantidade,
            'quantidadeHorizontal': self.quantidade_horizontal,
            'quantidadeVertical': self.quantidade_vertical,
            'margemSuperior': self.margem_superior,
            'margemInferior': self.margem_inferior,
            'margemEsquerda': self.margem_esquerda,
            'margemDireita': self.margem_direita,
            'espacamentoHorizontal': self.espacamento_horizontal,
            'espacamentoVertical': self.espacamento_vertical,
            'etiquetaTemplateModelList': [etiqueta_template_model.serialize() for etiqueta_template_model in self.etiqueta_template_model_list],
            'etiquetaFormatoPapelModel': self.etiqueta_formato_papel_model.serialize() if self.etiqueta_formato_papel_model else None,
        }